### PR TITLE
Custom time query

### PR DESCRIPTION
**Pull request by caoash**
https://github.com/varchar-io/nebula/issues/180

The git history and code are messy. I ran tests for creating timeline query by day and week since you can do those with constant bucket sizes. An issue that arises is that the table below the timeline will show a negative value for the initial time because the start of a day/week may be before beginTime.

To implement month/year (which needs consideration of leap years / 29, 30, 31 days), likely need to write a UDF. I'm looking into it but still a bit confused on how to do this. Any formatting / code structure feedback is appreciated.

### CODE REVIEW

**Comment by caoash** [comment]
Fixed the history with rebase

**Comment by caoash** [comment]
implemented the UDF that I want to apply to the ColumnExpression in `QueryHandler.cpp`, but I still can't figure out how to

File: src/api/udf/RoundToUnit.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        static constexpr int32_t MONTH_CASE = 3;
+        static constexpr int32_t QUARTER_CASE = 4;
+        static constexpr int32_t YEAR_CASE = 5;
+
+        // need to subtract begintime
+
+        InputType timeSecs = std::move(origin.value());
+        RoundedTimeType res = std::numeric_limits<RoundedTimeType>::lowest();
+        std::time_t timePoint = (std::time_t) timeSecs;
+        switch (unit) {
```
Comment (caoash) [inline src/api/udf/RoundToUnit.h:59]: likely can use #define here since most of the cases are the same except the function

File: src/common/Evidence.h [hunk 2]
```diff
@@ -147,7 +147,40 @@ class Evidence {
...
+    // yday stores [0-365]
+    auto delta_days = (tm->tm_yday) * 86400; 
+    return strip_days - delta_days;
+  }
+
+  // given a date time value, strip all but month
+  // such as value of (2019-05-15 23:01:34) => value of (2019-05-1 00:00:00)
+  static std::time_t month(std::time_t time) {
+    // remove all except for date
+    auto strip_days = date(time);
```
Comment (caoash) [inline src/common/Evidence.h:165]: is there some way to do this with move semantics since we don't need time anymore?

Comment (caoash) [reply src/common/Evidence.h:165]: i used to always just make copies, so not very clear on this kind of stuff

File: src/api/udf/RoundToUnit.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        static constexpr int32_t HOUR_CASE = 0;
+        static constexpr int32_t DAY_CASE = 1;
+        static constexpr int32_t WEEK_CASE = 2;
+        static constexpr int32_t MONTH_CASE = 3;
+        static constexpr int32_t QUARTER_CASE = 4;
+        static constexpr int32_t YEAR_CASE = 5;
+
+        // need to subtract begintime
+
+        InputType timeSecs = std::move(origin.value());
```
Comment (caoash) [inline src/api/udf/RoundToUnit.h:56]: is this safe to do?

Comment (caoash) [reply src/api/udf/RoundToUnit.h:56]: or even necessary?

Comment (shawncao) [reply src/api/udf/RoundToUnit.h:56]: No, we should not do this on const reference, also unnecessary since it's primitive data, there is no value but in fact worse in terms of perf - more CPU cycles.

File: src/api/udf/RoundToUnit.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * Define expressions used in the nebula DSL.
+ */
+namespace nebula {
+namespace api {
+namespace udf {
+// Round a time 
+using UdfRoundBase = nebula::surface::eval::UDF<nebula::type::Kind::BIGINT, nebula::type::Kind::BIGINT>;
+class RoundToUnit : public UdfRoundBase {
```
Comment (shawncao) [inline src/api/udf/RoundToUnit.h:32]: The UDF structure looks fine to me, but the name could be more concise, maybe `RoundTimeToUnit` or implicit "RoundTime"

File: src/common/Evidence.h [hunk 5]
```diff
@@ -147,7 +147,40 @@ class Evidence {
     return std::time(nullptr);
   }
 
-  // given a date time value, stripe time out to date granularity
+  // given a date time value, strip all but year
+  // such as value of (2019-05-15 23:01:34) => value of (2019-01-1 00:00:00)
+  static std::time_t year(std::time_t time) {
```
Comment (shawncao) [inline src/common/Evidence.h:152]: the logic seems correct.

File: src/common/Evidence.h [hunk 6]
```diff
@@ -147,7 +147,40 @@ class Evidence {
...
+    auto strip_days = date(time);
+    auto tm = std::gmtime(&strip_days);
+    // yday stores [0-365]
+    auto delta_days = (tm->tm_yday) * 86400; 
+    return strip_days - delta_days;
+  }
+
+  // given a date time value, strip all but month
+  // such as value of (2019-05-15 23:01:34) => value of (2019-05-1 00:00:00)
+  static std::time_t month(std::time_t time) {
```
Comment (shawncao) [inline src/common/Evidence.h:163]: you should write unit tests for all these new functions in CommonTests

File: src/common/Evidence.h (hunk 2 above)
Comment (shawncao) [reply src/common/Evidence.h:165]: this should be fine

**Comment by shawncao** [comment]
I see this is a good problem, thanks for the explanation, @caoash 

I like option 1 in your proposal, here are my thoughts:

Nebula as a backend engine/database, should not care how frontend will use it, but it has to simply let clients know how its behaviors are consistent, for this reason, we will rule out option 2.

For a similar reason, changing the client's query is "risky" as the client may not expect that, in fact, in many situations, different clients may expect different results, to solve this type of confusion, the best way is to not do anything, as long as your contract is simple enough, the client side should handle all the different situations.

So for this case, I think to keep the first time point to 0 is the simplest answer across all the queries, regardless of what type of time patterns they run with.

File: src/api/udf/RoundTimeToUnit.h [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        static constexpr int32_t DAY_CASE = 2;
+        static constexpr int32_t WEEK_CASE = 3;
+        static constexpr int32_t MONTH_CASE = 4;
+        static constexpr int32_t QUARTER_CASE = 5;
+        static constexpr int32_t YEAR_CASE = 6;
+
+        // need to subtract begintime
+
+        InputType timeSecs = origin.value();
+        RoundedTimeType res = std::numeric_limits<RoundedTimeType>::lowest();
```
Comment (shawncao) [inline src/api/udf/RoundTimeToUnit.h:58]: why not use the MIN_VALUE const rather than this function call? we use it all over other places.

File: src/api/udf/RoundTimeToUnit.h [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                res = (RoundedTimeType) roundedPoint;
+                break;
+            }
+            case QUARTER_CASE: {
+                break;
+            }
+            case YEAR_CASE: { 
+                std::time_t roundedPoint = nebula::common::Evidence::year(timePoint); 
+                // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+                res = (RoundedTimeType) roundedPoint;
```
Comment (shawncao) [inline src/api/udf/RoundTimeToUnit.h:91]: you can move this repeated line to line 95 before your assertion.

File: src/api/udf/RoundTimeToUnit.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                break;
+            }
+        }
+
+        N_ENSURE_NE(res, std::numeric_limits<RoundedTimeType>::lowest(), "res should be defined");
+        
+        // LOG(INFO) << "timeSecs: " << nebula::common::Evidence::fmt_normal(timeSecs);
+        // LOG(INFO) << "unit: " << unit;
+        // LOG(INFO) << "res: " << nebula::common::Evidence::fmt_normal(res);
+        // LOG(INFO) << "(res - beginTime): " << nebula::common::Evidence::fmt_normal(res - beginTime);
```
Comment (shawncao) [inline src/api/udf/RoundTimeToUnit.h:101]: clean up these lines

File: src/api/udf/RoundTimeToUnit.h [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+        RoundedTimeType minValue = (RoundedTimeType)(0);
+
+        return std::max(minValue, res - beginTime);
```
Comment (shawncao) [inline src/api/udf/RoundTimeToUnit.h:105]: can you add one line comment to make sure this is the contract that was agreed on - seems very important to make sure the value returned is positive.

File: src/api/udf/UDFFactory.h [hunk 11]
```diff
@@ -25,6 +25,7 @@
 #include "Max.h"
 #include "Min.h"
 #include "Not.h"
+#include "RoundTimeToUnit.h"
```
Comment (shawncao) [inline src/api/udf/UDFFactory.h:28]: sort includes by alpha order - move to line 30~31

File: src/api/udf/RoundTimeToUnit.h [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Define expressions used in the nebula DSL.
+ */
+namespace nebula {
+namespace api {
+namespace udf {
+// Round a time 
+using UdfRoundBase = nebula::surface::eval::UDF<nebula::type::Kind::BIGINT, nebula::type::Kind::BIGINT>;
+class RoundTimeToUnit : public UdfRoundBase {
```
Comment (shawncao) [inline src/api/udf/RoundTimeToUnit.h:32]: can we rename this class and file to `RoundTime.h`, I prefer short name if possible

File: src/common/Evidence.h [hunk 13]
```diff
@@ -147,7 +147,40 @@ class Evidence {
...
+    auto strip_days = date(time);
+    auto tm = std::gmtime(&strip_days);
+    // mday stores [1-31]
+    auto delta_days = (tm->tm_mday - 1) * 86400; 
+    return strip_days - delta_days;
+  }
+
+  // given a date time value, strip all but week
+  // such as value of (Wednesday 2019-05-15 23:01:34) => value of (Sunday 2019-05-12 00:00:00)
+  static std::time_t week(std::time_t time) {
```
Comment (shawncao) [inline src/common/Evidence.h:174]: can we support week day too, return 0~6 for [Sunday, Monday...Saturday] ?

File: src/common/Evidence.h [hunk 14]
```diff
@@ -147,7 +147,40 @@ class Evidence {
...
+  static std::time_t year(std::time_t time) {
+    // remove all except for date
+    auto strip_days = date(time);
+    auto tm = std::gmtime(&strip_days);
+    // yday stores [0-365]
+    auto delta_days = (tm->tm_yday) * 86400; 
+    return strip_days - delta_days;
+  }
+
+  // given a date time value, strip all but month
```
Comment (shawncao) [inline src/common/Evidence.h:161]: it will be great to support the quarter as well - Q1(Jan~Mar), Q2(April~Jun), Q3, Q4, a lot of financial data use these terms

File: src/common/Evidence.h (hunk 6 above)
Comment (shawncao) [reply src/common/Evidence.h:163]: if you have done this item, could you please reply and close it, so I don't need to track these open comments, we want to eventually close all the comments and then we will merge/

File: src/service/http/nebula/node.js [hunk 15]
```diff
@@ -228,6 +228,7 @@ const webq = (q, handler, client) => {
 
     // set query type and window
     req.setTimeline(state.timeline);
+    req.setTimeUnit(state.time_unit);
```
Comment (shawncao) [inline src/service/http/nebula/node.js:231]: cool - thanks for the end-to-end testing.

File: src/service/http/nebula/web.js [hunk 16]
```diff
@@ -303,6 +303,14 @@ const build = (s) => {
         keys.unshift(timeCol);
     }
 
+    let TIME_UNIT = 0;
+    let TIME_ARR = ["3600", "86400", "604800", "2592000", "QUARTERLY", "31536000"];
```
Comment (shawncao) [inline src/service/http/nebula/web.js:307]: is `QUARTERLY` expected to be like that?

File: src/service/http/nebula/web.js [hunk 17]
```diff
@@ -312,11 +320,14 @@ const build = (s) => {
         keys: keys,
         timeline: $('#timeline').is(':checked'),
         window: $$(windowId),
+        time_unit: TIME_UNIT.toString(),
         metrics: metrics,
         sort: $$('#ob'),
         limit: $$('#limit')
     };
 
+    console.log(JSON.stringify(state));
```
Comment (shawncao) [inline src/service/http/nebula/web.js:329]: remove it after testing

File: src/service/protos/nebula.proto [hunk 18]
```diff
@@ -208,6 +208,9 @@ message QueryRequest {
 
   // custom column list
   repeated CustomColumn custom = 12;
+
+  // custom time unit 
+  uint64 time_unit = 13;
```
Comment (shawncao) [inline src/service/protos/nebula.proto:213]: we can define a enum for this 
```
enum TimeUnit {
  HOUR = 1;
  DAY = 2;
  ....
}
```

File: src/service/server/QueryHandler.cpp [hunk 19]
```diff
@@ -152,10 +154,12 @@ inline SortType orderTypeConvert(OrderType type) {
 
 // build the query object to execute
 std::shared_ptr<Query> QueryHandler::build(const Table& tb, const QueryRequest& req, ErrorCode& err) const noexcept {
+  LOG(INFO) << "VALIDATING...";
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:157]: remove after testing

File: src/service/server/QueryHandler.cpp [hunk 20]
```diff
@@ -24,6 +24,8 @@
 #include "execution/core/ServerExecutor.h"
 #include "service/node/RemoteNodeConnector.h"
 
+#include "surface/MockSurface.h"
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:27]: why include this test utility in QueryHandler?

File: src/service/server/QueryHandler.cpp [hunk 21]
```diff
@@ -240,30 +244,40 @@ std::shared_ptr<Query> QueryHandler::buildQuery(const Table& tb, const QueryRequ
...
     N_ENSURE_GT(range, 0, "timeline requires end time greater than start time");
 
-    int32_t window = (int32_t)req.window();
+    int64_t window = (int64_t)req.window();
+    int64_t time_unit = (int64_t)req.time_unit();
+  
     auto buckets = window == 0 ? FLAGS_AUTO_WINDOW_SIZE : range / window;
+    LOG(INFO) << FLAGS_AUTO_WINDOW_SIZE;
+    LOG(INFO) << "number of buckets: " << buckets;
+    // LOG(INFO) << "window: " << window;
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:257]: again, make sure clean up all `LOG` lines

File: src/service/server/QueryHandler.cpp [hunk 22]
```diff
@@ -240,30 +244,40 @@ std::shared_ptr<Query> QueryHandler::buildQuery(const Table& tb, const QueryRequ
...
-
     // recalculate window based on buckets
     window = range / buckets;
     N_ENSURE_GT(window, 0, "window should be at least 1 second");
-
+    // LOG(INFO) << "recalculated window: " << window; 
     // only one bucket?
     std::shared_ptr<Expression> windowExpr = nullptr;
-    if (buckets < 2) {
+    int64_t beginTime = (int64_t)req.start();
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:268]: why moving this out of else block?

File: src/service/server/QueryHandler.cpp [hunk 23]
```diff
@@ -240,30 +244,40 @@ std::shared_ptr<Query> QueryHandler::buildQuery(const Table& tb, const QueryRequ
...
     }
-
     // recalculate window based on buckets
     window = range / buckets;
     N_ENSURE_GT(window, 0, "window should be at least 1 second");
-
+    // LOG(INFO) << "recalculated window: " << window; 
     // only one bucket?
     std::shared_ptr<Expression> windowExpr = nullptr;
-    if (buckets < 2) {
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:258]: clean up and revert all the formatting changes will help review without the noisy distractions.

File: src/storage/VectorReader.h [hunk 24]
```diff
@@ -87,14 +87,19 @@ class VectorRow : public nebula::surface::RowData {
     }
 
     auto& value = vector[realIndex];
-    if (value.IsNumber()) {
+    if (!value.IsString()) {
```
Comment (shawncao) [inline src/storage/VectorReader.h:90]: try to rebase to the latest, these changes seems unrelated.

File: src/surface/eval/EvalContext.h [hunk 25]
```diff
@@ -166,7 +166,9 @@ class TypeValueEval : public ValueEval {
 
 public:
   TypeValueEval(const std::string& sign, const ExpressionType et, const OPT&& op, const EvalBlock&& eb)
-    : TypeValueEval(sign, et, std::move(op), std::move(eb), {}, {}) {}
```
Comment (shawncao) [inline src/surface/eval/EvalContext.h:169]: revert

File: src/surface/eval/UDF.h [hunk 26]
```diff
@@ -125,6 +125,7 @@ enum class UDFType {
   PREFIX,
   IN,
   BETWEEN,
+  ROUNDTIMETOUNIT,
```
Comment (shawncao) [inline src/surface/eval/UDF.h:128]: let's call it short `ROUNDTIME`

File: src/surface/eval/UDF.h [hunk 27]
```diff
@@ -204,6 +205,11 @@ UDF_TRAITS(LIKE, nebula::type::Kind::BOOLEAN, nebula::type::Kind::VARCHAR)
 STATIC_TRAITS(PREFIX, false)
 UDF_TRAITS(PREFIX, nebula::type::Kind::BOOLEAN, nebula::type::Kind::VARCHAR)
 
+// define traits for UDF: ROUNDTIMETOUNIT
+// ROUNDTIMETOUNIT function applies to integer type only.
+STATIC_TRAITS(ROUNDTIMETOUNIT, false)
+UDF_TRAITS(ROUNDTIMETOUNIT, nebula::type::Kind::BIGINT, nebula::type::Kind::BIGINT)
```
Comment (shawncao) [inline src/surface/eval/UDF.h:211]: Looks good.

**Review by shawncao — COMMENTED** [review]
Just finished the review, I think overall this is the correct change and it is what we need for this feature, thanks for working on it and bring to real life. 
Congratulations, it is a really good job without much help from me!

@caosgc33d @johncao1 


I have reviewed each file, please make sure we don't have formatting changes (clang-format) and clean up the necessary debugging code. And then we can close all the comments, once we are done, we can merge it to master!

File: src/common/Evidence.h (hunk 6 above)
Comment (caoash) [reply src/common/Evidence.h:163]: done

**Comment by caoash** [comment]
I think rebase was successful

File: src/api/udf/In.h [hunk 28]
```diff
@@ -67,7 +66,6 @@ class In : public nebula::surface::eval::UDF<nebula::type::Kind::BOOLEAN, IK> {
         if (N_UNLIKELY(source == std::nullopt)) {
           return std::nullopt;
         }
-
```
Comment (shawncao) [inline src/api/udf/In.h:70]: as mentioned above, can you put back all the empty lines, please? in all the files, just quickly go through them.

File: src/api/udf/RoundTime.h [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+
+        static constexpr int32_t HOUR_CASE = 1;
+        static constexpr int32_t DAY_CASE = 2;
+        static constexpr int32_t WEEK_CASE = 3;
+        static constexpr int32_t MONTH_CASE = 4;
+        static constexpr int32_t QUARTER_CASE = 5;
+        static constexpr int32_t YEAR_CASE = 6;
+
+        auto timeSecs = origin.value();
+        LOG(INFO) << timeSecs;
```
Comment (shawncao) [inline src/api/udf/RoundTime.h:56]: remove

File: src/api/udf/RoundTime.h [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+          roundedPoint = nebula::common::Evidence::quarter(timePoint);
+          break;
+        }
+        case YEAR_CASE: {
+          roundedPoint = nebula::common::Evidence::year(timePoint);
+          break;
+        }
+        }
+        // LOG(INFO) << nebula::common::Evidence::fmt_normal(timeSecs);
+        auto form = nebula::common::Evidence::fmt_normal(roundedPoint);
```
Comment (shawncao) [inline src/api/udf/RoundTime.h:87]: not used, remove?

File: src/api/udf/RoundTime.h [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+          break;
+        }
+        case DAY_CASE: {
+          roundedPoint = nebula::common::Evidence::date(timePoint);
+          break;
+        }
+        case WEEK_CASE: {
+          roundedPoint = nebula::common::Evidence::week(timePoint);
+          break;
+        }
```
Comment (shawncao) [inline src/api/udf/RoundTime.h:72]: also add WeekDay case?

File: src/common/test/TestCommon.cpp [hunk 32]
```diff
@@ -303,6 +303,34 @@ TEST(CommonTest, TestTimeParsing) {
...
+    auto time1 = Evidence::time("2019-04-03 23:23:45", "%Y-%m-%d %H:%M:%S");
+    auto time2 = Evidence::time("2019-05-13 23:23:45", "%Y-%m-%d %H:%M:%S");
+
+    auto time_hour = Evidence::hour(time1);
+    auto time_hour_exp = Evidence::time("2019-04-03 23:00:00", "%Y-%m-%d %H:%M:%S");
+    EXPECT_EQ(time_hour, time_hour_exp);
+
+    auto time_week = Evidence::week(time1);
+    auto time_week_exp = Evidence::time("2019-03-31 00:00:00", "%Y-%m-%d %H:%M:%S");
+    EXPECT_EQ(time_week, time_week_exp);
```
Comment (shawncao) [inline src/common/test/TestCommon.cpp:318]: test weekday function too?

File: src/common/test/TestCommon.cpp [hunk 33]
```diff
@@ -303,6 +303,34 @@ TEST(CommonTest, TestTimeParsing) {
   }
 }
 
+TEST(CommonTest, TestStripTime) {
```
Comment (shawncao) [inline src/common/test/TestCommon.cpp:306]: The current tests are good, do you have more tests (especially edge cases) to add? These core functions deserve more coverage.

File: src/service/protos/nebula.proto (hunk 18 above)
Comment (shawncao) [reply src/service/protos/nebula.proto:213]: can you look at this again?

File: src/service/server/QueryHandler.cpp [hunk 34]
```diff
@@ -240,19 +240,19 @@ std::shared_ptr<Query> QueryHandler::buildQuery(const Table& tb, const QueryRequ
...
+
     auto buckets = window == 0 ? FLAGS_AUTO_WINDOW_SIZE : range / window;
     if (buckets == 0 || buckets > range) {
       buckets = range;
     }
-
     // recalculate window based on buckets
     window = range / buckets;
     N_ENSURE_GT(window, 0, "window should be at least 1 second");
-
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:255]: keep all the empty lines between code blocks.

File: src/service/server/QueryHandler.cpp [hunk 35]
```diff
@@ -285,6 +289,9 @@ std::shared_ptr<Query> QueryHandler::buildQuery(const Table& tb, const QueryRequ
     // group by clause uses 1-based index
     keys.push_back(columns.size());
   }
+  for (auto i = 0; i < (int)(keys.size()); i++) {
+    LOG(INFO) << keys[i];
+  }
```
Comment (shawncao) [inline src/service/server/QueryHandler.cpp:294]: remove the debug lines

**Review by shawncao — COMMENTED** [review]
Some more comments, getting closer. 

Please do not remove empty lines between code blocks.

**Review by shawncao — APPROVED** [review]
Just go through one time and do some cleanups (extra empty space, feel your clang-format is still not right), and then merge it into master!

Great job!!
